src/thermal.cpp

```diff
@@ -0,0 +1,289 @@
+#include "openmc/thermal.h"
+
+#include <algorithm> // for sort, move, min, max, find
+#include <cmath>     // for round, sqrt, abs
+#include <sstream>   // for stringstream
+
+#include "xtensor/xarray.hpp"
+#include "xtensor/xbuilder.hpp"
+#include "xtensor/xmath.hpp"
+#include "xtensor/xsort.hpp"
+#include "xtensor/xtensor.hpp"
+#include "xtensor/xview.hpp"
+
+#include "openmc/constants.h"
+#include "openmc/endf.h"
+#include "openmc/error.h"
+#include "openmc/random_lcg.h"
+#include "openmc/search.h"
+#include "openmc/secondary_correlated.h"
+#include "openmc/secondary_thermal.h"
+#include "openmc/settings.h"
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+namespace data {
+std::vector<std::unique_ptr<ThermalScattering>> thermal_scatt;
+std::unordered_map<std::string, int> thermal_scatt_map;
+}
+
+//==============================================================================
+// ThermalScattering implementation
+//==============================================================================
+
+ThermalScattering::ThermalScattering(hid_t group, const std::vector<double>& temperature)
+{
+  // Get name of table from group
+  name_ = object_name(group);
+
+  // Get rid of leading '/'
+  name_ = name_.substr(1);
+
+  read_attribute(group, "atomic_weight_ratio", awr_);
+  read_attribute(group, "energy_max", energy_max_);
+  read_attribute(group, "nuclides", nuclides_);
+
+  // Read temperatures
+  hid_t kT_group = open_group(group, "kTs");
+
+  // Determine temperatures available
+  auto dset_names = dataset_names(kT_group);
+  auto n = dset_names.size();
+  auto temps_available = xt::empty<double>({n});
+  for (int i = 0; i < dset_names.size(); ++i) {
+    // Read temperature value
+    double T;
+    read_dataset(kT_group, dset_names[i].data(), T);
```

Comment:
Question: It's slightly unusual to declare a primitive (`double` type here) then pass it to be modified to a function (`read_dataset`) by passing the primitive directly, rather than passing the pointer to the primitive.  I'm sure there's a reason you are doing this here, but I just want to double-check here and make sure that is correct.

Reply:
Agreed -- the interface to the `read_dataset` function is a bit awkward. In fact, the entire file it comes from needs a lot of clean up, so I've created an issue to track that (#1550)